File: dart-wrapper/cardinal_sdk/ios/Classes/CryptoPlugin.swift
import Flutter
import CardinalDartSdkSupportLib

class CryptoPlugin {

	public static func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
		let args = call.arguments as! [String: String]
		let methodName = call.method

		let completed = SubscriptionPlugin.dispatch(methodName: methodName, parameters: args) { success, errorCode, errorMessage, errorDetails in
			if (errorCode != nil){
				result(FlutterError(code: errorCode!, message: errorMessage, details: errorDetails))
			}
			else {
				result(success)
			}
		}

		if !completed {
			result(FlutterMethodNotImplemented)
		}
	}

	static func dispatch(
		methodName: String,
		parameters: [String : String],
		resultCallback: @escaping (
			String?,
			String?,
			String?,
			String?
		) -> Void
	) -> Bool {
		switch methodName {
			case "KeyPairRecoverer.waitForRecoveryKey": close(parameters: parameters, resultCallback: resultCallback)
			case "KeyPairRecoverer.recoverWithRecoveryKey": getCloseReason(parameters: parameters, resultCallback: resultCallback)
			default: return false
		}
		return true
	}

	private static func waitForRecoveryKey(
		parameters: [String : String],
		resultCallback: @escaping (
			String?,
			String?,
			String?,
			String?
		) -> Void
	) {
		KeyPairRecoverer.shared.waitForRecoveryKey(
			dartResultCallback: resultCallback,
            cancellationToken: Int64(parameters["cancellationToken"]!)!,
            referenceId: parameters["referenceId"]!,
            recoveryKeyString: parameters["recoveryKey"]!,
            autoDeleteString: parameters["autoDelete"]!,
            waitSecondsString: parameters["waitSeconds"]!
		)
	}

	private static func recoverWithRecoveryKey(
		parameters: [String : String],
		resultCallback: @escaping (
			String?,
			String?,
			String?,
			String?
		) -> Void
	) {
        KeyPairRecoverer.shared.recoverWithRecoveryKey(
			dartResultCallback: resultCallback,
            referenceId: parameters["referenceId"]!,
            recoveryKeyString: parameters["recoveryKey"]!,
            autoDeleteString: parameters["autoDelete"]!
		)
	}
}

File: dart-wrapper/cardinal_sdk/android/src/main/kotlin/com/icure/cardinal_sdk/CryptoPlugin.kt
package com.icure.cardinal_sdk

import com.icure.cardinal.sdk.dart.crypto.KeyPairRecoverer
import com.icure.cardinal.sdk.dart.utils.ApiScope
import com.icure.cardinal.sdk.dart.utils.Example
import com.icure.cardinal.sdk.dart.utils.ResourceUtils
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

object CryptoPlugin : MethodChannel.MethodCallHandler {

	@Suppress("UNCHECKED_CAST")
	override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
		if (
			!dispatch(
				methodName = call.method,
				parameters = (call.arguments as Map<String, String>?).orEmpty()
			) { success, errorCode, errorMessage, errorDetail ->
				if (errorCode != null)
					result.error(errorCode, errorMessage, errorDetail)
				else
					result.success(success)
			}
		) result.notImplemented()
	}

	fun dispatch(
		methodName: String,
		parameters: Map<String, String>,
		resultCallback: (
			String?,
			String?,
			String?,
			String?,
		) -> Unit,
	): Boolean = when(methodName) {
		"KeyPairRecoverer.recoverWithRecoveryKey" -> recoverWithRecoveryKey(parameters, resultCallback)
		"KeyPairRecoverer.waitForRecoveryKey" -> waitForRecoveryKey(parameters, resultCallback)
		else -> null
	}?.let { true } ?: false

	private fun recoverWithRecoveryKey(parameters: Map<String, String>, resultCallback: (
		String?,
		String?,
		String?,
		String?,
	) -> Unit) {
		KeyPairRecoverer.recoverWithRecoveryKey(
			resultCallback,
			referenceId = parameters.getValue("referenceId"),
			recoveryKeyString = parameters.getValue("recoveryKey"),
			autoDeleteString = parameters.getValue("autoDelete")
		)
	}

	private fun waitForRecoveryKey(parameters: Map<String, String>, resultCallback: (
		String?,
		String?,
		String?,
		String?,
	) -> Unit) {
		KeyPairRecoverer.waitForRecoveryKey(
			resultCallback,
			cancellationToken = parameters.getValue("cancellationToken").toLong(),
			referenceId = parameters.getValue("referenceId"),
			recoveryKeyString = parameters.getValue("recoveryKey"),
			autoDeleteString = parameters.getValue("autoDelete"),
			waitSecondsString = parameters.getValue("waitSeconds")
		)
	}
}